repair-missing/script.whizzml

```diff
@@ -0,0 +1,192 @@
+(define (flatten-dataset dataset-id)
+  (let (fds ((fetch dataset-id) "fields"))
+    (if (some (lambda (fid) (= ((fds fid) "optype") "datetime")) (keys fds))
+      (let (_ (log-info "Generated fields detected.  Flattening dataset...")
+            mod-id (create-model dataset-id)
+            bp (create-batchprediction {"dataset" dataset-id
+                                        "all_fields" true
+                                        "output_dataset" true
+                                        "model" (wait mod-id)
+                                        "prediction_name" "__bmldummy"})
+            pred-ds ((fetch (wait bp)) "output_dataset_resource")
+            new_fields ((fetch (wait pred-ds)) "fields")
+            pred_fid (id-from-fields new_fields "__bmldummy")
+            final-ds (wait (create-dataset {"origin_dataset" pred-ds
+                                            "excluded_fields" [pred_fid]})))
+        (delete* [mod-id pred-ds bp])
+        final-ds)
+      dataset-id)))
+
+;; Helper function to get the ID of a field given its name
+(define (id-from-fields fields name)
+  (loop (ids (keys fields) vs (values fields))
+    (cond (empty? ids) false
+          (= name ((head vs) "name" false)) (head ids)
+          (recur (tail ids) (tail vs)))))
+
+;; Resolve the objective field ID given the input.  Choose the default
+;; if it's empty but we still want to use an objective.
+(define (get-objective use-objective dataset-id input)
+  (let (fds ((fetch dataset-id) "fields"))
+    (cond (not use-objective) false
+          (empty? input) (dataset-get-objective-id dataset-id)
+          (contains? fds input) input
+          (let (oid (id-from-fields fds input))
+            (if oid
+              oid
+              (raise (str "Objective field '" input "' not found!")))))))
+
+;; Find fields that are missing data.  As a bonus, we'll find
+;; non-preferred fields and exclude them later.
+(define (get-bad-fields fds objective-field)
+  (let (exclude? (lambda (fid) (not (or ((fds fid) "preferred")
+                                        (= fid objective-field))))
+        to-exclude (sort (filter exclude? (keys fds)))
+        repair? (lambda (fid) (and (> (((fds fid) "summary") "missing_count") 0)
+                                   (not (= fid objective-field))
+                                   (not (member? fid to-exclude))))
+        to-repair (sort (filter repair? (keys fds))))
+    [to-repair to-exclude]))
+
+;; Compose a dataset containing only rows for which the specified
+;; field ID isn't missing
+(define (filter-missing dataset-id field-id)
+  (let (fexpr (str "(not (missing \"" field-id "\"))"))
+    (create-dataset {"origin_dataset" dataset-id "lisp_filter" fexpr})))
+
+;; Create a model with the given dataset for the given field ID if it
+;; is a modelable type, else return the empty string as a dummy
+;; resource ID.
+(define (make-model ds-name train-ds fid fdesc obj-id)
```

Comment:
That's what I was thinking, looking it up from the the `train-ds`. Which, now that you point it out, is a lot of calls. This is something I'm learning to pay attention to when coding, thanks for clarifying!